src/Tools/dotnet-monitor/ConfigurationJsonWriter.cs

```diff
@@ -63,12 +68,12 @@ public void Write(IConfiguration configuration, bool full, bool skipNotPresent)
 
                 //No sensitive information
                 ProcessChildSection(configuration, ConfigurationKeys.GlobalCounter, skipNotPresent, includeChildSections: true);
-                ProcessChildSection(configuration, ConfigurationKeys.CollectionRules, skipNotPresent, includeChildSections: true);
+                ProcessChildSection(configuration, ConfigurationKeys.CollectionRules, skipNotPresent, includeChildSections: true, sectionType: typeof(RootOptions).GetProperty(ConfigurationKeys.CollectionRules).PropertyType);
```

Comment:
Rather than attempting to understand the configuration with types, it might be easier to rely on the keys. If the children are 0,1,2,3,4 etc we show them as an array. Since we control the names, we don't have to worry that someone will create a config key with an actual name of 0 or something like that.

Reply:
That was my initial thought too, but is there anything keeping someone from naming their collection rule, "0"? I was concerned about clobbering things actually named as indices by guessing they were arrays

Reply:
I think we can detect that. I.e. if someone names their rules "0", "b", "5" we are not an array. If all the children are 0-n, we are an array. There is one edge cases we couldn't detect, and that is someone naming their rules '0', '1', '2', '3' etc. I honestly think it's ok if we show the wrong thing in that case rather than have to change this code every time we have add an array in configuration.

Reply:
That was the case I was worried about (or, in a more basic scenario, someone with only one rule that they just named "0" since they couldn't be bothered to choose a name). On the one hand, we would knowingly output illegal configuration, but that is the much simpler solution. I'll probably start playing around with this in another branch, but will wait until @jander-msft gets back from vacation and has a chance to weigh in before committing to one direction or another.

Reply:
This discussion also made me think of something else, which is that the `ConfigurationJsonWriter` uses colons to separate the different keys in the hierarchy (e.g. CollectionRules:abc:Actions:0). Technically, we allow colons in our collection rule names, which means someone can make the output illegally formatted if they put a colon in their collection rule name. That's probably a relatively minor bug, but I also don't see any harm in banning the inclusion of colons in keys unless someone is opposed.

Reply:
The configuration section will properly handle this case. You'll just see a Key with a ':' inside. This shouldn't be an issue.

Reply:
Are you sure? This is the behavior I see in `dotnet monitor` (the actual release from dotnet tools):

![ColonSeparationBug_ConfigShow](https://user-images.githubusercontent.com/85592574/142466922-77591cda-6e06-43cd-8daa-e7ce94fd6b36.png)
![ColonSeparationBug_CollectionRulesStart](https://user-images.githubusercontent.com/85592574/142466930-85920bdb-a62e-4c66-b992-b496247fea89.png)

Reply:
You're right. I tried a basic top level "a:b": "value" configuration and it worked. But when I changed it to "Nested":{ "a:b": "value"}, "a" and "b" became distinct elements. This looks like a bug in https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration.Json/src/JsonConfigurationFileParser.cs